src/databricks/labs/ucx/azure/locations.py

```diff
@@ -0,0 +1,170 @@
+import logging
+from urllib.parse import urlparse
+
+from databricks.labs.blueprint.installation import Installation
+from databricks.sdk import WorkspaceClient
+from databricks.sdk.errors.platform import PermissionDenied
+
+from databricks.labs.ucx.azure.access import AzureResourcePermissions
+from databricks.labs.ucx.azure.resources import AzureResources
+from databricks.labs.ucx.config import WorkspaceConfig
+from databricks.labs.ucx.framework.crawlers import StatementExecutionBackend
+from databricks.labs.ucx.hive_metastore import ExternalLocations
+
+logger = logging.getLogger(__name__)
+
+
+class ExternalLocationsMigration:
+    def __init__(
+        self,
+        ws: WorkspaceClient,
+        hms_locations: ExternalLocations,
+        resource_permissions: AzureResourcePermissions,
+        azurerm: AzureResources,
+    ):
+        self._ws = ws
+        self._hms_locations = hms_locations
+        self._resource_permissions = resource_permissions
+        self._azurerm = azurerm
+
+    @classmethod
+    def for_cli(cls, ws: WorkspaceClient, installation: Installation):
+        config = installation.load(WorkspaceConfig)
+        sql_backend = StatementExecutionBackend(ws, config.warehouse_id)
+        hms_locations = ExternalLocations(ws, sql_backend, config.inventory_database)
+        azurerm = AzureResources(ws)
+        resource_permissions = AzureResourcePermissions(installation, ws, azurerm, hms_locations)
+
+        return cls(ws, hms_locations, resource_permissions, azurerm)
+
+    def _app_id_credential_name_mapping(self) -> tuple[dict[str, str], dict[str, str]]:
+        # list all storage credentials.
+        # generate the managed identity/service principal application id to credential name mapping.
+        # return one mapping for all non read-only credentials and one mapping for all read-only credentials
+        # TODO: considering put this logic into the StorageCredentialManager
+        app_id_mapping_write = {}
+        app_id_mapping_read = {}
+        all_credentials = self._ws.storage_credentials.list(max_results=0)
+        for credential in all_credentials:
+            # cannot have none credential name, it's required for external location
+            if not credential.name:
+                continue
+            # if service principal based credential, use service principal's application_id directly
+            if credential.azure_service_principal:
+                if not credential.read_only:
+                    app_id_mapping_write[credential.azure_service_principal.application_id] = credential.name
+                    continue
+                if credential.read_only:
+                    app_id_mapping_read[credential.azure_service_principal.application_id] = credential.name
+            # if managed identity based credential, fetch the application_id of the managed identity
+            if credential.azure_managed_identity:
```

Comment:
Nit: This code can be de-nested with if-not-continue

Reply:
fixed.